### PR TITLE
Feature/reactive forms

### PR DESCRIPTION
**Pull request by wis-zadonsky**

### CODE REVIEW

File: src/app/auth/components/registration/registration.component.html [hunk 1]
```diff
@@ -0,0 +1,61 @@
+<section class="registration">
+  <div class="registration-container">
+
+    <pre>
+      {{ registrationForm.valid | json }}
+    </pre>
```
Comment (AntonGorelov) [inline src/app/auth/components/registration/registration.component.html]: Are u sure that u need to keep the debug data here?

File: src/app/auth/components/registration/registration.component.html [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    <form class="registration-form"
+          [formGroup]="registrationForm"
+          (ngSubmit)="registrationForm.valid && submit()">
+      <h1 class="mat-display-3">Registration</h1>
+
+      <mat-form-field hideRequiredMarker appearance="fill">
+        <mat-label>Email</mat-label>
```
Comment (AntonGorelov) [inline src/app/auth/components/registration/registration.component.html:9]: What is the point of using `hideRequiredMarker` directive here?

File: src/app/auth/components/registration/registration.component.html [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+      <mat-form-field hideRequiredMarker appearance="fill">
+        <mat-label>Confirm password</mat-label>
+        <input
+          type="password"
+          name="confirmPassword"
+          formControlName="confirmPassword"
+          matInput
+        >
+
+        <mat-error *ngIf="registrationForm.controls['confirmPassword']?.errors?.passwordMismatch">
```
Comment (AntonGorelov) [inline src/app/auth/components/registration/registration.component.html]: I guess will be better to make getter for control and put it in to the template. Obviously with using OnPush strategy.

File: src/app/auth/components/registration/registration.component.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+  @Output()
+  public formSubmit = new EventEmitter<IRegistration>();
+
+  public readonly registrationForm!: FormGroup;
+
+  constructor(
+    private readonly _formBuilder: FormBuilder,
+    private readonly _customValidator: CustomValidationService,
+  ) {
```
Comment (AntonGorelov) [inline src/app/auth/components/registration/registration.component.ts:27]: Are u sure that we need to create form twice?

File: src/app/auth/containers/login/login.container.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import { Component, OnInit } from '@angular/core';
+
+@Component({
+  selector: 'app-login-container',
+  templateUrl: './login.container.html',
+})
+export class LoginContainer implements OnInit {
```
Comment (AntonGorelov) [inline src/app/auth/containers/login/login.container.ts]: Lets remove unused hooks in the class?

File: src/app/auth/containers/registration/registration.container.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+@Component({
+  selector: 'app-registration-container',
+  templateUrl: './registration.container.html',
+})
+export class RegistrationContainer implements OnInit {
+
+  constructor(private readonly _authService: AuthService) { }
+
+  public ngOnInit(): void {
```
Comment (AntonGorelov) [inline src/app/auth/containers/registration/registration.container.ts]: Will be better to remove unused hooks in the class.

File: src/app/auth/directives/match-password.directive.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+import { CustomValidationService } from '../services/custom-validation.service';
+
+@Directive({
+  selector: '[appMatchPassword]',
+  providers: [{ provide: NG_VALIDATORS, useExisting: MatchPasswordDirective, multi: true }],
+})
+export class MatchPasswordDirective implements Validator {
+
+  @Input('appMatchPassword') public MatchPassword: string[] = [];
```
Comment (AntonGorelov) [inline src/app/auth/directives/match-password.directive.ts]: Why do u need to use array of string? And variable naming with capital letter doesn't match our style guide.

File: src/app/auth/services/auth.service.ts [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+import { IRegistration } from '../interfaces/registration.interface';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class AuthService {
+
+  constructor() { }
+
+  public signUp(formData: IRegistration): void {
```
Comment (AntonGorelov) [inline src/app/auth/services/auth.service.ts]: Are you going to keep personal data in local storage? It's ok just for training project.

File: src/app/books/models/book.model.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  public previews!: any;
+  public image!: string;
+  public writingDate!: Date;
+  public releaseDate!: Date;
+
+  constructor(book: any) {
+    this.id = book.id;
+    this.description = book.description || 'No description';
+    this.authorId = book.authorId;
+    this.author = book.author || 'No author';
```
Comment (AntonGorelov) [inline src/app/books/models/book.model.ts:22]: Incorrect default values! Could you check against the given types?

File: src/app/layout/header/header.component.html [hunk 10]
```diff
@@ -6,6 +6,7 @@ <h1 class="header__title">Book Shop</h1>
                 <li><a class="header__link" routerLink="/books">Books</a></li>
                 <li><a class="header__link" routerLink="/authors">Authors</a></li>
                 <li><a class="header__link" routerLink="/genres">Genres</a></li>
+                <li><a class="header__link" routerLink="/auth">Login</a></li>
             </ul>
```
Comment (AntonGorelov) [inline src/app/layout/header/header.component.html]: Don't forget to check against the page is inaccessible for authorised users.

File: src/styles.scss [hunk 11]
```diff
@@ -1,4 +1,10 @@
 /* You can add global styles to this file, and also import other style files */
 
 html, body { height: 100%; }
-body { font-family: Roboto, "Helvetica Neue", sans-serif; }
+body {
+  font-family: Roboto, "Helvetica Neue", sans-serif;
+
+  section {
```
Comment (AntonGorelov) [inline src/styles.scss]: Are u sure that we need to override global styles for it?

**Review by AntonGorelov — CHANGES_REQUESTED** [review]

File: src/styles.scss (hunk 11 above)
Comment (wis-zadonsky) [reply src/styles.scss]: Angular Material paste this by default

File: src/app/auth/components/registration/registration.component.html [hunk 12]
```diff
@@ -1,39 +1,50 @@
...
+          name="confirmPassword"
+          formControlName="confirmPassword"
           matInput
-          required
-          [(ngModel)]="formControls.repeatPassword"
-          #repeatPassword="ngModel"
         >
-        <mat-error *ngIf="repeatPassword.errors?.passwordMismatch">
-          Passwords mismatch
+        <mat-error *ngIf="confirmPasswordControl?.errors?.passwordMismatch">
```
Comment (AntonGorelov) [inline src/app/auth/components/registration/registration.component.html:46]: Will be better to check validation with `hasError()`.

File: src/app/auth/components/registration/registration.component.ts [hunk 13]
```diff
@@ -1,22 +1,44 @@
...
 
   @Output()
-  public formSubmit = new EventEmitter();
+  public formSubmit = new EventEmitter<IRegistration>();
+
+  public readonly registrationForm!: FormGroup;
 
-  public formControls: Partial<IRegistration> = {};
+  public get confirmPasswordControl(): AbstractControl {
+    return this.registrationForm.controls['confirmPassword'];
```
Comment (AntonGorelov) [inline src/app/auth/components/registration/registration.component.ts:21]: You forgot to improve this solution. Try to use the following code for getting AbstractControl:
`return this.registrationForm.get('confirmPassword');`

File: src/app/books/components/book-form/book-form.component.html [hunk 14]
```diff
@@ -2,20 +2,20 @@
   <div class="book-create-container">
     <h1 class="book-create__title">Create book</h1>
 
-    <form class="book-create-form" (ngSubmit)="submit()">
+    <form class="book-create-form" [formGroup]="createBookForm" (ngSubmit)="submit()">
       <mat-form-field appearance="legacy" hideRequiredMarker>
```
Comment (AntonGorelov) [inline src/app/books/components/book-form/book-form.component.html:6]: `hideRequiredMarker` is not a kind of suitable solution. Your control doesn't know that it's a required field.

File: src/app/books/models/book.model.ts [hunk 15]
```diff
@@ -18,7 +18,7 @@ export class BookModel {
   constructor(book: any) {
     this.id = book.id;
     this.description = book.description || 'No description';
-    this.authorId = book.authorId;
+    this.authorId = book.authorId || 'No author';
```
Comment (AntonGorelov) [inline src/app/books/models/book.model.ts:21]: I guess you forgot to fix this model.

**Pull request merged**